### PR TITLE
many: support console-conf as a snap

### PR DESCRIPTION
**Pull request by bboozzoo**
The branch adds support for console-conf as a snap. The console-conf-wrapper script is copied directly from https://github.com/canonical/subiquity/blob/0aaade5c53baf53e25dd4dfe8d003dd42cc395e8/bin/console-conf-wrapper with a tweak introduced in https://github.com/snapcore/core-base/pull/187/commits/e5ffa51c4a9c513a9366af9cfda17aad4deed91e so that we can properly detect when to redisplay the content.

The service definitions are also copied from the subiquity repository, with partial changes applied from https://github.com/canonical/subiquity/pull/1478. Unfortunately I had a problem with the branch when I applied it as is which could be observed as console-conf@ttyX stopping and getty@ttyX faling. I suspect it may be some sort of race. Anyways, in the current setup I can pretty consistently observe getty to start when needed.

When built without console-conf, the core24 snap is ~70MB, compared to 75MB currently reported by `snap info core24` for the upstream/published version.

### CODE REVIEW

File: Makefile
Comment (Meulengracht) [inline Makefile:39]: There shouldn't be anymore local debs, maybe we can remove this now?

File: hooks/001-extra-packages.chroot
Comment (Meulengracht) [inline hooks/001-extra-packages.chroot:88]: this too?

File: static/usr/lib/systemd/system/console-conf@.service
Comment (valentindavid) [inline static/usr/lib/systemd/system/console-conf@.service:20]: This is not right. We have to use conflicts. I had opened https://github.com/snapcore/core-base/pull/84/files, but then decided to push it upstream. Which was ignored https://github.com/canonical/subiquity/pull/1478.

If we have to have the static files here. We should remove this.

File: static/usr/lib/systemd/system/console-conf@.service
Comment (valentindavid) [inline static/usr/lib/systemd/system/console-conf@.service:23]: `OnSuccess=`?

File: static/usr/lib/systemd/system/console-conf@.service
Comment (valentindavid) [inline static/usr/lib/systemd/system/console-conf@.service:4]: Is this needed?

File: static/usr/lib/systemd/system/console-conf@.service
Comment (valentindavid) [inline static/usr/lib/systemd/system/console-conf@.service:7]: Remove `core18.start-snapd.service`?

File: static/usr/lib/systemd/system/console-conf@.service
Comment (bboozzoo) [reply static/usr/lib/systemd/system/console-conf@.service:20]: Let me double check that, with conflicts=getty@..I was not able to successfully transition from console-conf to getty on the first run of a UC console-conf-less image which ended up being console-conf & getty restarting in a loop

File: static/usr/lib/systemd/system/console-conf@.service
Comment (valentindavid) [inline static/usr/lib/systemd/system/console-conf@.service:15]: This needs an explanation I think. Original issue: https://bugs.launchpad.net/snappy/+bug/1650091

File: static/usr/lib/systemd/system/console-conf@.service
Comment (valentindavid) [inline static/usr/lib/systemd/system/console-conf@.service:32]: This needs an explanation. Systemd documentation recommends never using this. So it would be nice to explain why we need it in a comment.

File: static/usr/lib/systemd/system/console-conf@.service
Comment (valentindavid) [inline static/usr/lib/systemd/system/console-conf@.service:40]: Cleanup?

File: static/usr/lib/systemd/system/console-conf@.service
Comment (valentindavid) [inline static/usr/lib/systemd/system/console-conf@.service:11]: This also needs an explanation or be removed. I do not see why it should not be stopped when shutting down for instance.

Comment (bboozzoo) [reply static/usr/lib/systemd/system/console-conf@.service:11]: I didn't change too much of the service definitions wrt to what's in the upstream right here https://github.com/canonical/subiquity/blob/main/debian/console-conf.console-conf%40.service to keep the diff for a cross review small. If that's no needed I can clean this up, though I think it's be preferable to do it after the thing lands, anyways your call.

Comment (valentindavid) [reply static/usr/lib/systemd/system/console-conf@.service:11]: We can mark them as todos and test in a different PR.

File: static/usr/share/subiquity/console-conf-wrapper
Comment (valentindavid) [inline static/usr/share/subiquity/console-conf-wrapper:27]: Maybe we can switch this to use `/usr/libexec/core/get-mode`.

Comment (valentindavid) [reply static/usr/share/subiquity/console-conf-wrapper:27]: This can be done later.

File: static/usr/share/subiquity/console-conf-wrapper
Comment (valentindavid) [inline static/usr/share/subiquity/console-conf-wrapper:29]: And `/usr/libexec/core/get-arg` for this one.

File: static/usr/share/subiquity/console-conf-write-login-details
Comment (valentindavid) [inline static/usr/share/subiquity/console-conf-write-login-details:9]: I would quote `${iface}` in case it is empty.

File: static/usr/share/subiquity/console-conf-write-login-details
Comment (valentindavid) [inline static/usr/share/subiquity/console-conf-write-login-details:25]: `/home/*`?

Comment (bboozzoo) [reply static/usr/share/subiquity/console-conf-write-login-details:25]: yes, that's for the screen where user's login information is printed after boot

Comment (bboozzoo) [reply static/usr/share/subiquity/console-conf-write-login-details:25]: ah, nvm :)

Comment (valentindavid) [reply static/usr/share/subiquity/console-conf-write-login-details:25]: I feel like reading through `getent passwd`, then selecting any user with a shell that is not nologin, and get their home dir from there would be a bit better actually.

File: static/usr/share/subiquity/console-conf-write-login-details
Comment (valentindavid) [inline static/usr/share/subiquity/console-conf-write-login-details:35]: Why filtering out bridge interfaces? If the device is configured with bridge, then the hardware interfaces will have no ip addresses. And here we would get nothing. I would just show all IP from all interfaces that have an address.

File: static/usr/share/subiquity/console-conf-wrapper
Comment (Meulengracht) [inline static/usr/share/subiquity/console-conf-wrapper:95]: This looks like a typo, /var/ instead of /vat/

File: static/usr/share/subiquity/console-conf-write-login-details
Comment (valentindavid) [inline static/usr/share/subiquity/console-conf-write-login-details:25]: Remove `/root` and remove the `grep /home`?

File: static/usr/share/subiquity/console-conf-write-login-details
Comment (bboozzoo) [inline static/usr/share/subiquity/console-conf-write-login-details:8]: this needs to skip `lo`

File: static/usr/share/subiquity/console-conf-write-login-details
Comment (bboozzoo) [inline static/usr/share/subiquity/console-conf-write-login-details:32]: and this too

File: static/usr/lib/systemd/system/console-conf@.service
Comment (bboozzoo) [reply static/usr/lib/systemd/system/console-conf@.service:32]: I've dropped it

File: static/usr/share/subiquity/console-conf-wrapper
Comment (bboozzoo) [reply static/usr/share/subiquity/console-conf-wrapper]: done

File: static/usr/share/subiquity/console-conf-wrapper
Comment (bboozzoo) [reply static/usr/share/subiquity/console-conf-wrapper]: done too

File: static/usr/share/subiquity/console-conf-write-login-details
Comment (bboozzoo) [reply static/usr/share/subiquity/console-conf-write-login-details]: I think this was done by accident in the original code, when the idea was to skip `lo`. I've dropped this and added explicit check for `lo`.

File: static/usr/share/subiquity/console-conf-write-login-details
Comment (bboozzoo) [reply static/usr/share/subiquity/console-conf-write-login-details:25]: We actually have this weird entry in passwd which would show up otherwise:
```
sync:x:4:65534:sync:/bin:/bin/sync
```

Comment (valentindavid) [reply static/usr/share/subiquity/console-conf-write-login-details]: That is a deprecated user with no real use. We should remove that user.

Comment (bboozzoo) [reply static/usr/share/subiquity/console-conf-write-login-details]: Hm, I'll do that in a followup

File: static/usr/lib/systemd/system/serial-getty@.service.d/console-conf-serial.conf
Comment (bboozzoo) [reply static/usr/lib/systemd/system/serial-getty@.service.d/console-conf-serial.conf]: dropped now, the only downside I see is that gettys on tty2+ are started automatically when switching vts in qemu, which I don't think is a big deal anyway

File: Makefile
Comment (bboozzoo) [reply Makefile:39]: A followup maybe?

File: static/usr/lib/systemd/system/serial-getty@.service.d/console-conf-serial.conf
Comment (valentindavid) [reply static/usr/lib/systemd/system/serial-getty@.service.d/console-conf-serial.conf]: Thank you! I actually prefer that behavior.

File: static/usr/share/subiquity/console-conf-write-login-details
Comment (valentindavid) [reply static/usr/share/subiquity/console-conf-write-login-details]: Yep.

File: static/usr/lib/systemd/system/console-conf@.service
Comment (bboozzoo) [reply static/usr/lib/systemd/system/console-conf@.service]: dropped too

File: static/usr/lib/systemd/system/console-conf@.service
Comment (valentindavid) [reply static/usr/lib/systemd/system/console-conf@.service:11]: Can you add the link to the issue in a comment? So we can keep track later if we can remove by testing the case of the issue.

**Review by valentindavid — APPROVED** [review]
Looks good. There is still one of my comments to be addressed.

File: Makefile
Comment (Meulengracht) [reply Makefile:39]: Yep that is fine

File: static/usr/lib/systemd/system/console-conf@.service
Comment (Meulengracht) [reply static/usr/lib/systemd/system/console-conf@.service:11]: @bboozzoo

**Review by Meulengracht — APPROVED** [review]
lgtm

File: static/usr/lib/systemd/system/console-conf@.service
Comment (bboozzoo) [reply static/usr/lib/systemd/system/console-conf@.service:14]: this is done now, I've left a comment with explanation and a link to the bug report

**Pull request merged**
